File: frontend/src/api/pools.ts
import type { Pool, PoolDetail } from "../types/pool";

const BASE = "/api/pools";

async function json<T>(url: string, init?: RequestInit): Promise<T> {
  const res = await fetch(url, init);
  if (!res.ok) {
    throw new Error(`API error: ${res.status} ${res.statusText}`);
  }
  return res.json();
}

export async function createPool(
  name: string,
  description?: string
): Promise<{ id: number }> {
  return json(`${BASE}`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ name, description }),
  });
}

export async function createPoolWithStocks(
  name: string,
  ts_codes: string[],
  description?: string
): Promise<{ id: number; name: string; added: number; total_requested: number }> {
  return json(`${BASE}/batch`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ name, description, ts_codes }),
  });
}

export async function listPools(): Promise<Pool[]> {
  return json(`${BASE}`);
}

export async function getPoolDetail(poolId: number): Promise<PoolDetail> {
  return json(`${BASE}/${poolId}`);
}

export async function updatePool(
  poolId: number,
  data: { name?: string; description?: string }
): Promise<void> {
  await json(`${BASE}/${poolId}`, {
    method: "PUT",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(data),
  });
}

export async function deletePool(poolId: number): Promise<void> {
  await json(`${BASE}/${poolId}`, { method: "DELETE" });
}

export async function addStockToPool(
  poolId: number,
  tsCode: string
): Promise<void> {
  await json(`${BASE}/${poolId}/stocks`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ ts_code: tsCode }),
  });
}

export async function removeStockFromPool(
  poolId: number,
  tsCode: string
): Promise<void> {
  await json(`${BASE}/${poolId}/stocks/${tsCode}`, { method: "DELETE" });
}

File: backend/app/routers/pools.py
from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel
from sqlalchemy.ext.asyncio import AsyncSession

from app.db import get_db
from app.routers.quotes import get_manager
from app.datasources.manager import DataSourceManager
from app.services.pool_service import (
    add_stock_to_pool,
    create_pool,
    delete_pool,
    get_pool_detail,
    list_pools,
    remove_stock_from_pool,
    update_pool,
)

router = APIRouter(prefix="/api/pools", tags=["pools"])


class CreatePoolRequest(BaseModel):
    name: str
    description: str | None = None


class UpdatePoolRequest(BaseModel):
    name: str | None = None
    description: str | None = None


class CreateAndPopulateRequest(BaseModel):
    name: str
    description: str | None = None
    ts_codes: list[str]


class AddStockRequest(BaseModel):
    ts_code: str


@router.post("")
async def create(body: CreatePoolRequest, db: AsyncSession = Depends(get_db)):
    return await create_pool(db, body.name, body.description)


@router.post("/batch")
async def create_and_populate(
    body: CreateAndPopulateRequest,
    db: AsyncSession = Depends(get_db),
    manager: DataSourceManager = Depends(get_manager),
):
    """One-shot: create a pool and add all given ts_codes to it."""
    pool = await create_pool(db, body.name, body.description)
    pool_id = pool["id"]
    added = 0
    for ts in body.ts_codes:
        if await add_stock_to_pool(db, manager, pool_id, ts):
            added += 1
    return {"id": pool_id, "name": body.name, "added": added,
            "total_requested": len(body.ts_codes)}


@router.get("")
async def list_all(db: AsyncSession = Depends(get_db)):
    return await list_pools(db)


@router.get("/{pool_id}")
async def detail(pool_id: int, db: AsyncSession = Depends(get_db)):
    result = await get_pool_detail(db, pool_id)
    if result is None:
        raise HTTPException(status_code=404, detail="Pool not found")
    return result


@router.put("/{pool_id}")
async def update(pool_id: int, body: UpdatePoolRequest, db: AsyncSession = Depends(get_db)):
    result = await update_pool(db, pool_id, body.name, body.description)
    if result is None:
        raise HTTPException(status_code=404, detail="Pool not found")
    return result


@router.delete("/{pool_id}")
async def remove(pool_id: int, db: AsyncSession = Depends(get_db)):
    ok = await delete_pool(db, pool_id)
    if not ok:
        raise HTTPException(status_code=404, detail="Pool not found")
    return {"ok": True}


@router.post("/{pool_id}/stocks")
async def add_stock(
    pool_id: int,
    body: AddStockRequest,
    db: AsyncSession = Depends(get_db),
    manager: DataSourceManager = Depends(get_manager),
):
    ok = await add_stock_to_pool(db, manager, pool_id, body.ts_code)
    if not ok:
        raise HTTPException(status_code=404, detail="Pool not found")
    return {"ok": True}


@router.delete("/{pool_id}/stocks/{ts_code}")
async def remove_stock(
    pool_id: int, ts_code: str, db: AsyncSession = Depends(get_db)
):
    ok = await remove_stock_from_pool(db, pool_id, ts_code)
    if not ok:
        raise HTTPException(status_code=404, detail="Stock not in pool")
    return {"ok": True}
